src/Common/Crypto/OpenSSLInitializer.cpp

```diff
@@ -79,6 +79,8 @@ OpenSSLInitializer::~OpenSSLInitializer()
             chassert(OSSL_PROVIDER_unload(default_provider));
             default_provider = nullptr;
         }
+
+        OPENSSL_cleanup();
```

Comment:
Maybe it is a theoretical risk but the refcount in OpenSSLInitializer may (in theory) go up and down in arbitrary ways, in particular ~OpenSSLInitializer may be called > 1 times.

`OPENSSL_cleanup()` may be called just once during the process lifetime (--> https://github.com/ClickHouse/ClickHouse/pull/80138/files#r2086425898).

So it would be slightly cleaner to call this function during the actual server shutdown and not here.

Reply:
Hmm, let's make a `void cleanup()` with a separate counter which only goes up, and call `OPENSSL_cleanup()` from it only if `counter==0`.

Reply:
Also yes, let's try to call it on shutdown explicitly.